### DISCUSSION TITLE
Confused.com

### DISCUSSION THREAD

**Discussion by paul-uz** [type-challenges/type-challenges#12718] [General]
I found this repo from a YouTube video and it piqued my interest as I've started using TS at work.

I'm confused as to what these Types are for exactly and how they would be used in the real world?

A lot of them just seem like a function, so why not write them as functions?

Why would I need a type that "does" something? I thought types just describe what a value is eg a string or number? 

Hope someone can help me wrap my head around this!

**Comment by zigante** [comment]
In Typescript, types (as well as interfaces) are for exactly what you said about describing what something is, but not just for variable values, but also for more complex statements like functions and classes. The advantage of using Typescript is in the complex types that you can develop yourself.

<br/>

For simple cases like you said, we can rewrite the native types and rename them:
```ts
type MagicNumber = number;
type SuperString = string;
```

<br/>

But now, let's think about a variable called `status`. This variable will store the value of a transaction coming from the database. What is the best approach with a simple validation to see if the status is cancelled?

Using native type             |  Using custom type
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/42415920/177865036-67a99d5e-88df-4ba7-97cd-54058568392b.gif)  |  ![](https://user-images.githubusercontent.com/42415920/177865040-57440f82-cf11-4fd3-ad33-6b786bbe62a8.gif)

This is the simplest use of a type in Typescript. We can go even further and create a type that describes a function
```ts
type MyFunc = (value: number) => boolean
```
This is a widely used approach mainly in implementing the well-known clean code best practices. When implementing the Interface Segregation principle from  S.O.L.I.D, for example, we need to ensure that a specific function within one class must have the same signature as another function in another distinct class.

<br/>

Going even further, we can create conditional types:
In this example, let's imagine that we have an API that can respond to success or error, for that I will create two different signatures, one for error and one for success:
```ts
type SuccessResponse = { success: boolean; message: string };
type ErrorResponse = SuccessResponse & { error: Error };
```
And then I create the signature of a function that will return different objects depending on the input value:
![](https://user-images.githubusercontent.com/42415920/177869599-1912b57f-0067-43ea-adac-9395ac7032ff.gif)


### <h3 align="center" id="heading">Remembering that this is just the tip of the iceberg which is the wonderful world of Typescript.</h3>
I hope I was helpfull. If so, please mark the answer as accepted, if not, please let me know and let's continue our delicious discussion. :)

**Reply by paul-uz** [reply]
Thank you for taking the time to reply.

I do understand everything you mentioned. But I'm still confused about these challenges. 

Why are the challenges to create a Type and not just a function?

Can these Types, once discovered,  be used as a function or are they simply to describe a function? 

If they are to describe a function, and you also have to write the actual function in order to do that "operation", aren't you just doubling the amount of code you need to write?

I'm struggling to understand the benefits of writing a Type that describes a function.

**Reply by zigante** [reply]
The type is simply the signature (description) of a function, so much so that you can say what parameters it receives and its return type.  note that there is no implementation inside the function. Just the return value. 

The benefit are these: being able to reuse the types in different functions and also facilitating the use of the created function (not the creation of the function, but the created function usage)
